ui/tests/integration/components/kv/kv-patch-editor-test.js

```diff
@@ -0,0 +1,438 @@
+    assert.dom(FORM.patchUndo(2)).hasText('Remove', 'Undo button reads "Remove" for new keys');
+
+    // assert a new row is added
+    this.assertEmptyRow(assert);
+  });
+
+  test('it renders loading state', async function (assert) {
+    this.isSaving = true;
+    await this.renderComponent();
+
+    assert.dom(FORM.saveBtn).isDisabled();
+    assert.dom(FORM.cancelBtn).isDisabled();
+    assert.dom(`${FORM.saveBtn} ${GENERAL.icon('loading')}`).exists();
+  });
+
+  module('it submits', function () {
+    test('patch data for existing, deleted and new keys', async function (assert) {
+      await this.renderComponent();
+
+      // patch existing key
+      await click(FORM.patchEdit(0));
+      await fillIn(FORM.valueInput(0), 'bar');
+      // in qunit we have to unfocus the input so the following click event works on first try
+      await blur(FORM.valueInput(0));
+
+      // delete existing key
+      await click(FORM.patchDelete(1));
+      assert.dom(FORM.patchAlert('delete', 1)).hasText('This key value pair is marked for deletion.');
+      assert.dom(FORM.keyInput(1)).hasStyle({ 'text-decoration': 'line-through rgb(59, 61, 69)' });
+      assert.dom(`${FORM.patchAlert('delete', 1)} ${GENERAL.icon('trash')}`).exists();
+      // value is set to null under the hood, confirm the non-string warning doesn't display
+      assert
+        .dom(FORM.patchAlert('value-warning', 1))
+        .doesNotExist('non-string warning does not render for null values');
+
+      // add new key and click add
+      await fillIn(FORM.keyInput('new'), 'aKey');
+      await fillIn(FORM.valueInput('new'), 'aValue');
+      await click(FORM.patchAdd);
+
+      // add new key and do NOT click add
+      await fillIn(FORM.keyInput('new'), 'bKey');
+      await fillIn(FORM.valueInput('new'), 'bValue');
+
+      await click(FORM.saveBtn);
+
+      const [data] = this.onSubmit.lastCall.args;
+      assert.propEqual(
+        data,
+        { baz: null, foo: 'bar', aKey: 'aValue', bKey: 'bValue' },
+        `onSubmit called with ${JSON.stringify(data)}`
+      );
+    });
+
+    test('patch data when every action is canceled', async function (assert) {
+      await this.renderComponent();
+
+      await click(FORM.patchEdit(0));
+      await fillIn(FORM.valueInput(0), 'bar');
+      // in qunit we have to unfocus the input so the following click event works on the first try
+      await blur(FORM.valueInput(0));
+      await click(FORM.patchDelete(1));
+
+      await fillIn(FORM.keyInput('new'), 'aKey');
+      await fillIn(FORM.valueInput('new'), 'aValue');
+      await click(FORM.patchAdd);
+
+      // undo every action
+      await click(FORM.patchUndo(0)); // undo edit
+      await click(FORM.patchUndo(1)); // undo delete
+      await click(FORM.patchUndo(2)); // remove new row
+      await click(FORM.saveBtn);
+
+      const [data] = this.onSubmit.lastCall.args;
+      assert.propEqual(data, {}, `onSubmit called with ${JSON.stringify(data)}`);
```

Comment:
If i'm understanding this test, we are making sure that undoing the changes calls the changes with an empty object. Should we instead not call `onSubmit` and treat it as a cancel?

Reply:
I was going to ask the team about this...I wasn't sure if the parent should handle it? I feel like it makes sense to keep the logic in the child so the parent just calls the patch adapter method. Calling the @onCancel function seems reasonable, that seems like intuitive UX that no changes takes you back to the overview page 🤔

Reply:
If the parent should handle not actually calling the API if the data is empty? 🤔 Could go either way I suppose. A good topic for discussion for the team